db/chained_delegating_reader.hh

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/shared_future.hh>
+
+#include "flat_mutation_reader.hh"
+
+// A reader which allows to insert a deferring operation before reading.
+// All calls will first wait for a future to resolve, then forward to a given underlying reader.
+class chained_delegating_reader : public flat_mutation_reader::impl {
+    std::unique_ptr<flat_mutation_reader> _underlying;
+    std::function<future<flat_mutation_reader>()> _populate_reader;
+    shared_future<> _ready;
+    std::function<void()> _on_destroyed;
+    bool _populating_started;
+    
+public:
+    chained_delegating_reader(schema_ptr s, std::function<future<flat_mutation_reader>()> populate, reader_permit permit, std::function<void()> on_destroyed = []{})
+        : impl(s, std::move(permit))
+        , _populate_reader(std::move(populate))
+        , _on_destroyed(std::move(on_destroyed))
+        , _populating_started(false)
+    { }
+
+    chained_delegating_reader(chained_delegating_reader&& rd) = delete;
+
+    ~chained_delegating_reader() {
+        _on_destroyed();
+    }
+
+    virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+        if (!_populating_started) {
```

Comment:
_populating_started is not really necessary, because flat_mutation_reader doesn't support concurrent invocations. You can assume that all invocations are serialized by the user.

Reply:
The condition could be `if (!_underlying) {`